source/common/ssl/context_config_impl.h

```diff
@@ -33,17 +33,28 @@ class ContextConfigImpl : public virtual Ssl::ContextConfig {
                ? nullptr
                : certficate_validation_context_provider_->secret();
   }
+  const TrustedCaConfig* trustedCa() const override {
+    if (trusted_ca_provider_ != nullptr) {
+      return trusted_ca_provider_->secret();
+    } else if (certficate_validation_context_provider_ != nullptr &&
+               certficate_validation_context_provider_->secret() != nullptr) {
+      return certficate_validation_context_provider_->secret()->trustedCa();
+    }
+    return nullptr;
+  }
   unsigned minProtocolVersion() const override { return min_protocol_version_; };
   unsigned maxProtocolVersion() const override { return max_protocol_version_; };
 
   bool isReady() const override {
     // Either tls_certficate_provider_ is nullptr or
     // tls_certficate_provider_->secret() is NOT nullptr and
     // either certficate_validation_context_provider_ is nullptr or
-    // certficate_validation_context_provider_->secret() is NOT nullptr.
+    // certficate_validation_context_provider_->secret() is NOT nullptr and
+    // either trusted_ca_provider_ is nullptr or trusted_ca_provider_->secret() is NOT nullptr.
     return (!tls_certficate_provider_ || tls_certficate_provider_->secret() != nullptr) &&
            (!certficate_validation_context_provider_ ||
-            certficate_validation_context_provider_->secret() != nullptr);
+            certficate_validation_context_provider_->secret() != nullptr) &&
+           (!trusted_ca_provider_ || trusted_ca_provider_->secret() != nullptr);
```

Comment:
This is not easy to read.  How about split it into two calls.   isTlsCertificateReady() and isValidationContextReady().  
IsRead() has to be both ready.